core/claims/claim.go

```diff
@@ -131,35 +111,187 @@ func NewClaimFromEntry(e *merkletree.Entry) (merkletree.Entrier, error) {
+	// identity found in the value part of the claim.
+	ClaimRecipIdenValue ClaimRecip = 0b10
+)
+
+// ClaimHeader represents the first bytes of the claim index and contains its
+// type and flags.
+type ClaimHeader struct {
+	Type       ClaimType
+	Dest       ClaimRecip
+	Expiration bool
+	Version    bool
+}
+
+func bool2byte(b bool) byte {
+	if b {
+		return 1
+	}
+	return 0
+}
+
+func byte2bool(b byte) bool {
+	return b != 0
+}
+
+// Marshal the ClaimHeader into an entry
+func (c *ClaimHeader) Marshal(e *merkletree.Entry) {
+	index := e.Index()
+	copy(index[0][:ClaimTypeLen], c.Type[:])
+	flags0 := &index[0][ClaimTypeLen]
+	*flags0 = 0
+	*flags0 |= byte(c.Dest)
+	*flags0 |= bool2byte(c.Expiration) << 2
+	*flags0 |= bool2byte(c.Version) << 3
+}
+
+// Unmarshal the ClaimHeader from an entry
+func (c *ClaimHeader) Unmarshal(e *merkletree.Entry) {
+	index := e.Index()
+	copy(c.Type[:], index[0][:ClaimTypeLen])
+	flags0 := index[0][ClaimTypeLen]
+	c.Dest = ClaimRecip(flags0 & 0b00000011)
+	c.Expiration = byte2bool(flags0 & (1 << 2))
+	c.Version = byte2bool(flags0 & (1 << 3))
+}
+
+var (
+	// ClaimHeaderBasic is a simple claim type that can be used for anything.
+	ClaimHeaderBasic = ClaimHeader{
+		Type:       *ClaimTypeBasic,
+		Dest:       ClaimRecipSelf,
+		Expiration: false,
+		Version:    false}
+	// ClaimTypeAuthorizeKSignBabyJub is a claim type to autorize a babyjub public key for signing.
+	ClaimHeaderAuthorizeKSignBabyJub = ClaimHeader{
+		Type:       *ClaimTypeAuthorizeKSignBabyJub,
+		Dest:       ClaimRecipSelf,
+		Expiration: false,
+		Version:    false}
+)
+
+// EntrierMetadata is an intefrace that extends Entrier with a function that
+// returns the claim metadata.
+type EntrierMetadata interface {
+	merkletree.Entrier
+	Metadata() *Metadata
+}
+
+// Metadata is a header and generic (some optional) values of a claim.
+type Metadata struct {
+	header     ClaimHeader
+	Dest       *core.ID
+	Expiration int64
+	Version    uint32
+	RevNonce   uint32
+}
+
+// NewMetadata creates a new Metadata with a specific header.
+func NewMetadata(header ClaimHeader) Metadata {
+	return Metadata{header: header}
+}
+
+// Header returns the header from the metadata.
+func (m *Metadata) Header() ClaimHeader {
+	return m.header
+}
+
+// Type returns the claim type from the header in the metadata.
+func (m *Metadata) Type() ClaimType {
+	return m.header.Type
+}
+
+// Marshal the Metadata into an entry
+func (m *Metadata) Marshal(e *merkletree.Entry) {
+	m.header.Marshal(e)
+	index := e.Index()
+	value := e.Value()
+	switch m.header.Dest {
+	case ClaimRecipSelf:
+	case ClaimRecipIdenIndex:
+		copy(index[1][:], m.Dest[:])
+	case ClaimRecipIdenValue:
+		copy(value[1][:], m.Dest[:])
+	default:
+		panic(fmt.Sprintf("Unexpected header.Dest %v", m.header.Dest))
+	}
+	if m.header.Version {
+		binary.BigEndian.PutUint32(index[0][ClaimTypeLen+ClaimFlagsLen:], m.Version)
+	}
+	if m.header.Expiration {
+		binary.BigEndian.PutUint64(value[0][32/8:], uint64(m.Expiration))
```

Comment:
Why ClaimExpirationLen is a hardcoded value, while ClaimTypeLen,ClaimVersionLen, ClaimFlagsLen, ClaimFlagsLen, ClaimHeaderLen, ClaimVersionLen, ClaimRevNonceLen are declared as constants? In this way when we need to have a claim that needs Expiration, we have the constant of the len to do the parser.
With the ClaimExpirationLen or not, maybe here should use `value[0][ClaimRevNonceLen:]`?

Reply:
I have modified the code to use the `ClaimRevNonceLen` constant